### PR TITLE
add functions for scraping data from data.un.org tables

### PR DESCRIPTION
**Pull request by chacalle**
## Describe changes

Initial functions to scrape data from any of these undata tables http://data.un.org/Explorer.aspx

Current workflows that this helps with:
- manual download and prep of data from [un dyb yearbooks](https://unstats.un.org/unsd/demographic-social/products/dyb/). Need to confirm that these tables have all the same data. 
- manual filter, download & prep 100,000 rows at a time from each of the tables we need. This is our current workflow for getting dyb census data from http://data.un.org/Data.aspx?d=POP&f=tableCode%3a22.
- manual download & prep of wpp estimates

Future considerations:
- does the [undata api](http://data.un.org/Host.aspx?Content=API) provide access to the tables we need. Also this [link](http://www.undata-api.org/)
- we often need to download all the data in these tables. These new functions are purposefully throttled so that we don't overwhelm the site but I think it will be pretty slow. Can the api or using [RSelenium](https://docs.ropensci.org/RSelenium/index.html) to "click" the download button speed up entire downloads?

### Packages Repositories

* [X] Have you read the [contributing guidelines](https://github.com/ihmeuw-demographics/packageTemplate/wiki#guide-to-r-package-development) for `ihmeuw-demographics` R packages?
* [X] Have you successfully run `devtools::check()` locally?
* [X] Have you updated or added function (and vignette if applicable) documentation? Did you update the 'man' and 'NAMESPACE' files with `devtools::document()`?
* [X] Have you added in tests for the changes included in the PR?
* [X] Do the changes follow the `ihmeuw-demographics` [code style](https://github.com/ihmeuw-demographics/packageTemplate/wiki/Code-style-guide)?
* [ ] Do the changes need to be immediately included in a new build of [`docker-base`](https://github.com/ihmeuw-demographics/docker-base) or [`docker-internal`](https://github.com/ihmeuw-demographics/docker-internal)? If so follow directions in those repositories to rebuild and redeploy the images.
* [ ] Do the changes require updates to other repositories which use this package? If yes, make the necessary updates in those repos, and consider integration tests for those repositories.

### CODE REVIEW

**Comment by chacalle** [comment]
Are there other composite websites that it would be useful to scrape from? Can we use jenkins or github actions to automatically check for updates and scrape the websites when needed?

**Comment by hcomfo95** [comment]
> Are there other composite websites that it would be useful to scrape from? Can we use jenkins or github actions to automatically check for updates and scrape the websites when needed?

I think you could do something similar for HFD?

Main website: https://www.humanfertility.org/cgi-bin/main.php
Updates listed here: https://www.humanfertility.org/cgi-bin/news.php

For each country (under detailed data OR preliminary release), we currently click on the country name, click on "Input Data" (fourth tab at the top), and then click on years in the first row under Input Data corresponding to the Births [dataset.]

hfd_example

**Comment by hcomfo95** [comment]
> Are there other composite websites that it would be useful to scrape from? Can we use jenkins or github actions to automatically check for updates and scrape the websites when needed?

For HFC: 
Main website: https://www.fertilitydata.org/cgi-bin/index.php
Updates: https://www.fertilitydata.org/cgi-bin/news.php

Download the underlined file:
hfc_example

File: R/get_undata.R [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+#'
+#' @inheritParams polite::scrape
+#' @param pages \[`numeric()`\]\cr
+#'   Pages to scrape from the table. See `get_undata_npages` to scrape total
+#'   number of tables. Default is 1.
+#'
+#' @return \[`tibble()`\] all pages from the table combined together.
+#'
+#' @details
+#' A list of all possible tables is shown here http://data.un.org/Explorer.aspx.
```
Comment (krpaulson) [inline R/get_undata.R:14]: Can you include here or in the examples section a sampling of tables the demographics team might want to scrape?

File: R/get_undata.R [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+#' @title Scrape data.un.org tables
+#'
+#' @description Scrape multiple pages from any data.un.org table.
+#'
+#' @inheritParams polite::scrape
+#' @param pages \[`numeric()`\]\cr
+#'   Pages to scrape from the table. See `get_undata_npages` to scrape total
+#'   number of tables. Default is 1.
+#'
+#' @return \[`tibble()`\] all pages from the table combined together.
```
Comment (krpaulson) [inline R/get_undata.R]: Is this confusing given that most of our other functions return data tables?

File: R/get_undata.R [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+#'   other related undata functions.
+#'
+#' @family get_undata
+#'
+#' @examples
+#' table_url <- "http://data.un.org/Data.aspx?d=PopDiv&f=variableID%3a12"
+#' query <- undata_url_to_query(table_url)
+#'
+#' @export
+undata_url_to_query <- function(table_url, base_url = "http://data.un.org/Data.aspx") {
```
Comment (krpaulson) [inline R/get_undata.R:160]: I appreciate that these are mostly very small functions. Makes it easier to follow. A vignette might be helpful though.

**Review by krpaulson — DISMISSED** [review]

File: R/get_undata.R (hunk 3 above)
Comment (chacalle) [reply R/get_undata.R:160]: Yes for sure, will create a new issue for the vignette.

File: R/get_undata.R (hunk 1 above)
Comment (chacalle) [reply R/get_undata.R:14]: Good idea, will include this in the vignette issue

File: R/get_undata.R (hunk 2 above)
Comment (chacalle) [reply R/get_undata.R]: Yeah initially I was thinking it was good to just return unmodified from `rvest` but I agree now that it is more important to stay consistent for our team members who expect `data.table` currently.

**Comment by chacalle** [comment]
> > Are there other composite websites that it would be useful to scrape from? Can we use jenkins or github actions to automatically check for updates and scrape the websites when needed?
> 
> I think you could do something similar for HFD?
> 
> Main website: https://www.humanfertility.org/cgi-bin/main.php
> Updates listed here: https://www.humanfertility.org/cgi-bin/news.php
> 
> For each country (under detailed data OR preliminary release), we currently click on the country name, click on "Input Data" (fourth tab at the top), and then click on years in the first row under Input Data corresponding to the Births [dataset.]
> 
> hfd_example

Thanks for the examples @hcomfo95! This reminded me of the [`HMDHFDplus` package](https://cran.r-project.org/web/packages/HMDHFDplus/HMDHFDplus.pdf). I'll make an issue for us to look into the details of that package more.

**Pull request merged**